osv/ecosystems/ecosystems_test.py

```diff
@@ -0,0 +1,58 @@
+# Copyright 2024 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Ecosystem helpers tests."""
+
+import json
+
+import re
+
+import unittest
+
+from . import _ecosystems
+
+
+class EcosystemsTest(unittest.TestCase):
+  """Ecosystem helpers tests."""
+
+  def setUp(self):
+    with open('osv/osv-schema/validation/schema.json') as schema_f:
+      schema = json.load(schema_f)
+    self.schema_ecosystems = re.compile(schema['$defs']['ecosystem']['pattern'])
+    self.canonical_ecosystems = _ecosystems._ecosystems.keys()  # pylint: disable=protected-access
+
+  def test_ecosystem_supported_by_schema(self):
+    """Test ecosystems referenced exist in schema definition"""
+    for ecosystem in self.canonical_ecosystems:
+      self.assertTrue(
+          self.schema_ecosystems.match(ecosystem),
+          msg=f'"{ecosystem}" not defined in schema')
```

Comment:
Wondering if there might be a way to also check the inverse of this.
i.e. check if there's any ecosystems that are allowed in the schema that `ecosystems.get()` doesn't have.

Though, I can't really think of a way to do this given it's a regex pattern (unless we want to fuzz all possible strings)

Reply:
> Wondering if there might be a way to also check the inverse of this.
> i.e. check if there's any ecosystems that are allowed in the schema that `ecosystems.get()` doesn't have.

I had thought of that, but then it's somewhat to be expected as new ecosystems onboard, that they get defined in the schema first and then in OSV.dev code, so we wouldn't want this test to start failing when that happens, right?

Reply:
I suppose that makes sense.
On the other hand, the tests failing would be good to signal that we need to add the ecosystem before we start attempting to ingest them into the database.

Reply:
> On the other hand, the tests failing would be good to signal that we need to add the ecosystem before we start attempting to ingest them into the database.

That makes me think there might be an opportunity to flag a premature addition to sources{_test}.yaml 🤔

Reply:
No, that isn't going to work because there's no direct linkage between a source and the ecosystems it provides...